src/sagemaker/jumpstart/notebook_utils.py

```diff
@@ -12,9 +12,395 @@
+        )
+        if model_specs.training_supported:
+            scripts.add(JumpStartScriptScope.TRAINING)
+
+        if scripts == {e.value for e in JumpStartScriptScope}:
+            break
+    return sorted(list(scripts))
+
+
+def list_jumpstart_models(  # pylint: disable=redefined-builtin
+    filter: Union[Operator, str] = Constant(BooleanValues.TRUE),
+    region: str = JUMPSTART_DEFAULT_REGION_NAME,
+    list_incomplete_models: bool = False,
+    list_old_models: bool = False,
+    list_versions: bool = False,
+) -> List[Union[Tuple[str], Tuple[str, str]]]:
+    """List models for JumpStart, and optionally apply filters to result.
+
+    Args:
+        filter (Union[Operator, str]): Optional. The filter to apply to list models. This can be
+            either an ``Operator`` type filter (e.g. ``And("task == ic", "framework == pytorch")``),
+            or simply a string filter which will get serialized into an Identity filter.
+            (e.g. ``"task == ic"``). If this argument is not supplied, all models will be listed.
+            (Default: Constant(BooleanValues.TRUE)).
+        region (str): Optional. The AWS region from which to retrieve JumpStart metadata regarding
+            models. (Default: JUMPSTART_DEFAULT_REGION_NAME).
+        list_incomplete_models (bool): Optional. If a model does not contain metadata fields
+            requested by the filter, and the filter cannot be resolved to a include/not include,
+            whether the model should be included. By default, these models are omitted from results.
+            (Default: False).
+        list_old_models (bool): Optional. If there are older versions of a model, whether the older
+            versions should be included in the returned result. (Default: False).
+        list_versions (bool): Optional. True if versions for models should be returned in addition
+            to the id of the model. (Default: False).
+    """
+
+    model_id_version_dict: Dict[str, List[str]] = dict()
+    for model_id, version in _generate_jumpstart_model_versions(
+        filter=filter, region=region, list_incomplete_models=list_incomplete_models
+    ):
+        if model_id not in model_id_version_dict:
+            model_id_version_dict[model_id] = list()
+        model_id_version_dict[model_id].append(Version(version))
+
+    if not list_versions:
+        return sorted(list(model_id_version_dict.keys()))
+
+    if not list_old_models:
+        model_id_version_dict = {
+            model_id: set([max(versions)]) for model_id, versions in model_id_version_dict.items()
+        }
+
+    model_id_version_set: Set[Tuple[str, str]] = set()
+    for model_id in model_id_version_dict:
+        for version in model_id_version_dict[model_id]:
+            model_id_version_set.add((model_id, str(version)))
+
+    return sorted(list(model_id_version_set), key=cmp_to_key(_compare_model_version_tuples))
+
+
+def _generate_jumpstart_model_versions(  # pylint: disable=redefined-builtin
```

Comment:
nit: This functione is very large can we make this more modular?

Reply:
Thanks for the feedback. I agree this is a relatively long function. Would you mind if I create a backlog item to improve this logic?